csharp/ql/lib/semmle/code/csharp/dataflow/SSA.qll

```diff
@@ -566,6 +581,54 @@ module Ssa {
     override Location getLocation() { result = this.getCallable().getLocation() }
   }
 
+  private module NearestLocationInput implements NearestLocationInputSig {
+    class C = ImplicitParameterDefinition;
+
+    predicate relevantLocations(ImplicitParameterDefinition def, Location l1, Location l2) {
+      not def.getBasicBlock() instanceof ControlFlow::BasicBlocks::EntryBlock and
+      l1 = def.getParameter().getALocation() and
+      l2 = def.getBasicBlock().getLocation()
+    }
+  }
+
+  pragma[nomagic]
+  private predicate implicitEntryDef(ImplicitEntryDefinition def, SourceVariable v, Callable c) {
+    v = def.getSourceVariable() and
+    c = def.getCallable()
+  }
+
+  /**
+   * An SSA definition representing the implicit initialization of a parameter
+   * at the beginning of a callable.
+   */
+  class ImplicitParameterDefinition extends ImplicitEntryDefinition {
+    private Parameter p;
+
+    ImplicitParameterDefinition() {
+      exists(SourceVariable sv, Callable c |
+        implicitEntryDef(this, sv, c) and
+        localScopeSourceVariable(sv, p, _, c)
+      )
+    }
+
+    /** Gets the parameter that this entry definition represents. */
+    Parameter getParameter() { result = p }
+
+    override Element getElement() { result = this.getParameter() }
+
+    override string toString() {
+      result = SsaImpl::getToStringPrefix(this) + "SSA param(" + this.getParameter() + ")"
+    }
+
+    override Location getLocation() {
+      not NearestLocation<NearestLocationInput>::nearestLocation(this, _, _) and
```

Comment:
Why doesn't this give us the exact location whe the is only a single implementation?

Reply:
Because the
```
not def.getBasicBlock() instanceof ControlFlow::BasicBlocks::EntryBlock
```
constraint in `NearestLocationInput::relevantLocations/3` restricts the logic to multi-bodied methods.